code/parse/sexp_container.cpp

```diff
@@ -329,6 +337,233 @@ sexp_container *get_sexp_container_special(const SCP_string& text, size_t start_
+			break;
+
+		case ListModifier::REMOVE_LAST:
+			replacement_text = list_data.back();
+			list_data.pop_back();
+			break;
+
+		case ListModifier::GET_RANDOM:
+			data_index = util::Random::next((int)list_data.size());
+			replacement_text = *std::next(list_it, data_index);
+			break;
+
+		case ListModifier::REMOVE_RANDOM:
+			data_index = util::Random::next((int)list_data.size());
+			std::advance(list_it, data_index);
+			replacement_text = *list_it;
+			list_data.erase(list_it);
+			break;
+
+		case ListModifier::AT_INDEX:
+			number_string = text.substr(lookHere + modifier_length);
+			Assert(!number_string.empty());
+			data_index = atoi(number_string.c_str());
+			if (data_index < 0 || data_index >= (int)list_data.size()) {
+				Warning(LOCATION,
+					"sexp_container_replace_refs_with_values() found invalid index '%s' into list container %s",
+					number_string.c_str(),
+					container.container_name.c_str());
+				return false;
+			}
+			replacement_text = *std::next(list_it, data_index);
+
+			// include spaces, in case the FREDder put one between "At" and the index
+			number_length = strspn(number_string.c_str(), " 0123456789");
+			break;
+
+		default:
+			Warning(LOCATION, "sexp_container_replace_refs_with_values() found a container called %s to replace but the modifer is not recognised", container.container_name.c_str());
+			return false;
+		}
+
+		num_chars_to_replace += modifier_length;
+
+		// for the "At" modifier, we need to include the chars used by the index
+		if (modifier == ListModifier::AT_INDEX) {
+			num_chars_to_replace += number_length;
+		}
+	}
+
+	// account for the final '&'
+	++num_chars_to_replace;
+
+	return true;
+}
+
+// inspired by sexp_replace_variable_names_with_values()
+/**
+* Replace container references in a string with their values
+**/
+bool sexp_container_replace_refs_with_values(SCP_string &text)
+{
+	bool replaced_anything = false;
+	size_t lookHere = 0;
+	int num_replacements = 0;
+
+	while (lookHere < text.length()) {
+		// look for the meta-character
+		size_t foundHere = text.find(sexp_container::DELIM, lookHere);
+
+		// found?
+		if (foundHere != SCP_string::npos && foundHere != text.length() - 1) {
+			// see if a container starts at the next char
+			auto *p_container = get_sexp_container_special(text, foundHere + 1);
+			if (p_container != nullptr) {
+				// we want to replace either &list_container&list_modifier& or &map_container&map_key&
+
+				// + 2 for the enclosing '&' chars around the container name
+				size_t num_chars_to_replace = p_container->container_name.length() + 2;
+
+				// make sure the enclosing '&' appears after the container name
+				size_t expected_amp_pos = foundHere + num_chars_to_replace - 1;
+				if (expected_amp_pos >= text.length() || text[expected_amp_pos] != sexp_container::DELIM) {
+					// no modifier/key specified - ignore this matching container name, as the replacement is malformed
+					lookHere = foundHere + 1;
+					Warning(LOCATION,
```

Comment:
Again, should be an Assertion here, since the mission save function will add that last `&`?

Reply:
Got it.